[indexer-alt-framework] merge regulator and broadcaster

emmazzz opened the pull request:
## Description 

This PR merges regulator and broadcaster into one task so that checkpoint data is directly broadcast to the downstream pipelines.

## Test plan 
Modified existing tests.

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] gRPC:
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

amnn commented on crates/sui-indexer-alt-framework/src/ingestion/client.rs:
```diff
@@ -102,6 +102,13 @@ impl IngestionClient {
     }
 
+    #[cfg(test)]
+    pub(crate) fn new_for_testing(client: Arc<dyn IngestionClientTrait>) -> Self {
```
The things that are specific to testing in this function are to do with metrics. It would be better to introduce a `new_for_testing` on `IndexerMetrics` for this.

emmazzz commented:
Thanks for the review @amnn! I've addressed the comments in the latest commit so that there is a ranged broadcaster task that's handled by the loop, and spawned ad-hoc. This should make the grpc streaming integration easier so that we just need to add an outer loop that creates streaming and/or broadcaster handle (with peeked streaming cp as the broadcast end), and add to the inner loop one tokio select branch handling streaming.

emmazzz commented on crates/sui-indexer-alt-framework/src/ingestion/broadcaster.rs:
```diff
@@ -1,83 +1,583 @@
+        let (ingest_hi_watch_tx, ingest_hi_watch_rx) = watch::channel(initial_ingest_hi);
+
+        // Spawn the broadcaster task.
+        let mut broadcaster_handle = tokio::spawn(broadcast_range(
```
It will be at the top of the outer loop when we have an outer loop for streaming. But right now there's only the inner loop for `tokio select`ing cancellation, ingest_hi receiving, etc.

emmazzz commented on crates/sui-indexer-alt-framework/src/ingestion/broadcaster.rs:
```diff
@@ -1,83 +1,583 @@
+                    match result {
+                        Ok(Ok(())) => {
+                            info!("Broadcaster completed successfully");
+                            break;
```
> change the loop into a while that checks whether there are any checkpoints left to process after this current round has completed.

In this PR the current round is the only round since `broadcast_range` covers the entire range so we are sure that once the task completes successfully the work is done. With subscription there will be an outer while loop doing this check.

emmazzz commented on crates/sui-indexer-alt-framework/src/ingestion/broadcaster.rs:
```diff
@@ -1,83 +1,583 @@
+            );
+        }
+
+        assert_eq!(
```
Yup! This assert is redundant and can be removed.
> there's a possibility that the channel could witness some other checkpoints after having seen the expected checkpoints, and we wouldn't check that.

If we want to verify that we don't receive any more checkpoints, we have the `expect_timeout` function. Sometimes we need to do this checks as checkpoints come in to test the backpressure logic though.

amnn replied:
Using an expectation of timeout would be a worse way to perform this check because you need to wait long enough to be sure things have stabilised in the happy case (so test times increase) as opposed to telling the system to wind down and then performing your tests.
